### ISSUE TITLE
[Bug] UpdateWeightFromDistributed fails for VLM models in non-colocate mode

### ISSUE THREAD

**Issue by ZHZisZZ** [THUDM/slime#1673]
### Bug Description

When running VLM (Vision-Language Model) training without `--colocate`, the weight update from training to SGLang produces corrupted weights, causing SGLang to crash on the first `generate` request after weight sync.

The root cause is that `UpdateWeightFromDistributed` uses the manual `convert_to_hf()` function (regex-based name conversion) instead of the megatron-bridge for weight conversion. The bridge correctly handles all VLM parameters (vision encoder + language model), while the manual converter does not produce correct weights for the SGLang `load_weights()` path.

This is **not** an async-specific issue. Any non-colocate VLM configuration will trigger it, because:
- `--colocate` → `UpdateWeightFromTensor` → uses `HfWeightIteratorBridge` → correct
- no `--colocate` → `UpdateWeightFromDistributed` → uses `convert_to_hf()` → crash

Text-only models (e.g., Qwen3-4B) work fine in both modes because their `convert_to_hf()` implementation is correct.

### Steps to Reproduce

Run the following script which is a non-colocate version of `examples/geo3k_vlm/run_geo3k_vlm.sh`:

```bash
bash examples/geo3k_vlm/run_geo3k_vlm_non_colocate.sh
```

The only difference from the working `run_geo3k_vlm.sh` is replacing:
```bash
# Colocate (working):
MISC_ARGS=(
   --colocate
)
--actor-num-gpus-per-node ${NUM_GPUS}
```
with:
```bash
# Non-colocate (crashes):
MISC_ARGS=(
   --actor-num-nodes 1
   --actor-num-gpus-per-node $((NUM_GPUS / 2))
   --rollout-num-gpus $((NUM_GPUS / 2))
)
```

The crash happens at the initial `actor_model.update_weights()` call, before the first rollout.

### Expected Behavior

Non-colocate VLM training should work the same as colocate mode — weight updates should produce valid weights that SGLang can use for inference.

### Actual Behavior

TODO

### Environment

- slime version: latest (commit 1e99f92)
- Python version: 3.12
- PyTorch version: 2.x
- CUDA version: 12.x
- GPU type and count: 8x H100
- OS: Linux
- SGLang version: latest
- Megatron-LM version: latest

### Logs

```shell
Training side completes weight update successfully (Qwen3-VL-8B-Instruct example):

[2026-03-07 00:07:00] timer.py:24 - Timer update_weights start
[2026-03-07 00:07:00] memory_utils.py:41 - [Rank 0] Memory-Usage before update_weights: {'gpu': '0', 'total_GB': 79.19, 'free_GB': 50.83, 'used_GB': 28.36}
[2026-03-07 00:07:03] memory_utils.py:41 - [Rank 0] Memory-Usage after update_weights: {'gpu': '0', 'total_GB': 79.19, 'free_GB': 46.43, 'used_GB': 32.76}
[2026-03-07 00:07:03] timer.py:32 - Timer update_weights end (elapsed: 3.5s)


SGLang side receives the update and acknowledges:

POST /update_weights_from_distributed HTTP/1.1 200 OK  (multiple calls, all 200)
POST /continue_generation HTTP/1.1 200 OK


Then all generate requests fail:

POST /generate HTTP/1.1 500 Internal Server Error  (repeated)
```

### Verified working configurations
| Configuration | Weight update class | Conversion method | Result |
|---|---|---|---|
| VLM + colocate | `UpdateWeightFromTensor` | `HfWeightIteratorBridge` | Works |
| Text + non-colocate | `UpdateWeightFromDistributed` | `convert_qwen2_to_hf()` | Works |
| **VLM + non-colocate (8B-Instruct)** | **`UpdateWeightFromDistributed`** | **`convert_qwen3vl_to_hf()`** | **500 error** |
| **VLM + non-colocate (4B-Thinking)** | **`UpdateWeightFromDistributed`** | **`convert_qwen3vl_to_hf()`** | **NaN crash** |

### Key code paths

The branching point is in `slime/backends/megatron_utils/actor.py`:
```python
update_weight_cls = UpdateWeightFromTensor if self.args.colocate else UpdateWeightFromDistributed
```

- `UpdateWeightFromTensor` (colocate): uses `HfWeightIteratorBridge` which leverages `megatron-bridge` for correct architecture-aware weight conversion
- `UpdateWeightFromDistributed` (non-colocate): uses `convert_to_hf()` → `convert_qwen3vl_to_hf()`, a manual regex-based converter

Note that `UpdateWeightFromTensor` can also broadcast weights via NCCL to overflow engines (see `_send_hf_params()`), and this works because it sends **bridge-converted** HF weights. The NCCL broadcast mechanism itself is fine — the issue is purely in the weight conversion.

### Potential fix direction

Make `UpdateWeightFromDistributed` use `HfWeightIteratorBridge` (or equivalent bridge-based conversion) instead of the manual `convert_to_hf()` for VLM models.

### Pre-submission Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/THUDM/slime/blob/main/CONTRIBUTING.md) and understand the collaboration scope.
- [x] I have read the [documentation](https://thudm.github.io/slime/) and my issue is not addressed there.
- [x] I have searched for [existing issues](https://github.com/THUDM/slime/issues) and this is not a duplicate.
- [x] I have provided a minimal, reproducible example.

**Comment by ZHZisZZ** [comment]
@cjy0x I have not solved the issue yet; I closed it because I felt the descriptions were not clear enough. I will create another issue later. But If you would love to, I would also encourage you to create an issue too! Thanks

**Comment by coding-famer** [comment]
@cjy0x @ZHZisZZ Hi both~ I have a pr #1537 for this. You may give it a try

**Comment by ZHZisZZ** [comment]
@coding-famer Thanks!

**Comment by cjy0x** [comment]
> [@cjy0x](https://github.com/cjy0x) [@ZHZisZZ](https://github.com/ZHZisZZ) Hi both~ I have a pr [#1537](https://github.com/THUDM/slime/pull/1537) for this. You may give it a try

Hi, thanks for the PR. I tested this on `Qwen3-VL-30B-Instruct`, but I noticed the `raw_reward` doesn't improve in either the **colocated** or **separated** configurations. In fact, the reward actually trended downward over 250 steps of RL training. Have you tested this specific model on Slime before, or do you have any advice on what might be causing this? @coding-famer

**Comment by coding-famer** [comment]
> > [@cjy0x](https://github.com/cjy0x) [@ZHZisZZ](https://github.com/ZHZisZZ) Hi both~ I have a pr [#1537](https://github.com/THUDM/slime/pull/1537) for this. You may give it a try
> 
> Hi, thanks for the PR. I tested this on `Qwen3-VL-30B-Instruct`, but I noticed the `raw_reward` doesn't improve in either the **colocated** or **separated** configurations. In fact, the reward actually trended downward over 250 steps of RL training. Have you tested this specific model on Slime before, or do you have any advice on what might be causing this? [@coding-famer](https://github.com/coding-famer)

Did you add these arguments to `provider` for moe model?
https://github.com/THUDM/slime/blob/243d0883b90e407ee56a6a51d2d8d80fe5cae47e/examples/geo3k_vlm/README.md?plain=1#L15-L20

**Comment by coding-famer** [comment]
> > > > > [@cjy0x](https://github.com/cjy0x) [@ZHZisZZ](https://github.com/ZHZisZZ) Hi both~ I have a pr [#1537](https://github.com/THUDM/slime/pull/1537) for this. You may give it a try
> > > > 
> > > > 
> > > > Hi, thanks for the PR. I tested this on `Qwen3-VL-30B-Instruct`, but I noticed the `raw_reward` doesn't improve in either the **colocated** or **separated** configurations. In fact, the reward actually trended downward over 250 steps of RL training. Have you tested this specific model on Slime before, or do you have any advice on what might be causing this? [@coding-famer](https://github.com/coding-famer)
> > > 
> > > 
> > > Did you add these arguments to `provider` for moe model?
> > > [slime/examples/geo3k_vlm/README.md](https://github.com/THUDM/slime/blob/243d0883b90e407ee56a6a51d2d8d80fe5cae47e/examples/geo3k_vlm/README.md?plain=1#L15-L20)
> > > Lines 15 to 20 in [243d088](/THUDM/slime/commit/243d0883b90e407ee56a6a51d2d8d80fe5cae47e)
> > > **Important:** We use [Megatron Bridge](https://github.com/NVIDIA-NeMo/Megatron-Bridge) to support multimodal models. However, not all Megatron arguments are passed through to Megatron Bridge — you may need to set some manually [here](https://github.com/THUDM/slime/blob/de84e10d468dcb726e1199fd6bd16aa9538aed09/slime/backends/megatron_utils/model_provider.py#L89) (currently only parallelization-related arguments are passed). For example, for Qwen3-VL-30B-A3B you may need to add:
> > > provider.moe_aux_loss_coeff = args.moe_aux_loss_coeff
> > > provider.freeze_language_model = False
> > > provider.freeze_vision_model = False
> > 
> > 
> > I didnt realize this, thank you for your advice! but I run multi turn instead of single turn, do I still need these arguments (I didnt find them in multi turn's readme file) ? Anyway, I will give it a try. [@coding-famer](https://github.com/coding-famer)
> 
> Sorry to bother you. I added these three arguments in `model_provider.py` and tested again, but the curve stays the same. Does Slime support qwen3-vl-30b **multi-turn** or not? [@coding-famer](https://github.com/coding-famer)
> 
> Here are my configurations on H100:
> 
> rollout_args = (
>     f"--prompt-data {TRAIN_DATA_PATH} "
>     "--input-key problem "
>     "--label-key answer "
>     '--multimodal-keys \'{"image": "images"}\' '
>     "--rm-type math "
>     "--apply-chat-template "
>     "--custom-generate-function-path examples.geo3k_vlm_multi_turn.rollout.generate "
>     "--custom-config-path examples/geo3k_vlm_multi_turn/geo3k_vlm_multi_turn_config.yaml "
>     "--rollout-shuffle "
>     "--num-rollout 3000 "
>     "--rollout-batch-size 32 "
>     "--n-samples-per-prompt 8 "
>     "--rollout-max-response-len 4096 "
>     "--rollout-temperature 1 "
>     "--global-batch-size 256 "
> )
> 
> grpo_args = (
>     "--advantage-estimator grpo "
>     "--kl-loss-coef 0.00 "
>     "--kl-loss-type low_var_kl "
>     "--kl-coef 0.00 "
>     "--entropy-coef 0.00 "
>     "--eps-clip 0.2 "
>     "--eps-clip-high 0.28 "
>     "--use-kl-loss "
> )
> 
> optimizer_args = (
>     "--optimizer adam "
>     "--lr 1e-6 "
>     "--lr-decay-style constant "
>     "--weight-decay 0.1 "
>     "--adam-beta1 0.9 "
>     "--adam-beta2 0.98 "
> 
>     "--optimizer-cpu-offload "
>     "--overlap-cpu-optimizer-d2h-h2d "
>     "--use-precision-aware-optimizer "
> )
> 
> sglang_args = (
>     "--rollout-num-gpus-per-engine 4 "
>     "--sglang-mem-fraction-static 0.6 "
>     f"--sglang-cuda-graph-bs {' '.join(map(str, [1, 2, 4, 8] + list(range(16, 257, 8))))} "
> )
> 
> megatron_args = (
>     "--train-backend megatron "
>     f"--load /data/{MODEL_NAME} "
>     f"--ref-load /data/{MODEL_NAME} "
>     "--tensor-model-parallel-size 2 "
>     "--sequence-parallel "
>     "--pipeline-model-parallel-size 1 "
>     "--context-parallel-size 1 "
>     "--expert-model-parallel-size 8 "
>     "--expert-tensor-parallel-size 1 "
>     "--recompute-granularity full "
>     "--recompute-method uniform "
>     "--recompute-num-layers 1 "
>     "--use-dynamic-batch-size "
>     "--max-tokens-per-gpu 16384 "
>     "--balance-data "
>     "--attention-dropout 0.0 "
>     "--hidden-dropout 0.0 "
>     "--accumulate-allreduce-grads-in-fp32 "
>     "--attention-softmax-in-fp32 "
>     "--attention-backend flash "
>     "--megatron-to-hf-mode bridge "
> )
> 
> misc_args = (
>     "--actor-num-nodes 1 "
>     f"--actor-num-gpus-per-node 4 "
>     f"--rollout-num-gpus 4 " 
>     # f"--colocate "
> )
> and here‘s the raw_reward curve
> 
> <img alt="Image" width="1312" height="776" src="https://private-user-images.githubusercontent.com/239078838/569713689-f83e3c90-28c1-48cd-b5a7-730d431d4df1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ1Mjg3MTUsIm5iZiI6MTc3NDUyODQxNSwicGF0aCI6Ii8yMzkwNzg4MzgvNTY5NzEzNjg5LWY4M2UzYzkwLTI4YzEtNDhjZC1iNWE3LTczMGQ0MzFkNGRmMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMyNlQxMjMzMzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lOTA4ZGQzODJlYzVkNzg0MGJjNjEwNmRkNjQ1NTg1NDJlNTllM2JhMmViNzA5YmYzNjllYTliYTk2ZGJkNTcwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.fN1_GEeOckg5axyaCSbikIJQ7Vyc3GXemW9pwAAGeUQ">

I have tested qwen3-vl-moe multiturn in colocate mode, it works well.

**Comment by coding-famer** [comment]
@cjy0x 
Sorry, I'm not a maintainer of this repo and don't have enough computing resources to test it.

Could you first verify whether the issue is:

+ MoE-related
+ Multi-turn-related
+ Disaggregation-related

(I recall you mentioned that both the colocated and disaggregated modes have the same issue?)

I believe all of these are orthogonal. I'm also happy to help narrow things down with more details.

**Comment by huaiyizhao** [comment]
I wonder if using fsdp as backend solves the problem
